File: 06-DevOps/02-CI-CD/Jenkins声明式流水线/python/jenkins_pipeline.py
"""Jenkins 声明式 Pipeline 语义模型(可执行)。

权威依据: jenkins.io «Pipeline Syntax»(doc/book/pipeline/syntax/)。

本 demo 只实现**光读文档看不出来、必须实测/精读才能确定**的那几处规则:

1. **agent 与 timeout 的计数区间不同**(官方明确):
   - 顶层 ``agent``: 先分配 agent, 再应用 ``timeout`` ⇒ **分配 agent 的时间不计入** timeout;
   - stage 级 ``agent``: ``options`` 在**分配 agent 之前**被调用 ⇒ **分配时间计入** timeout,
     因此 agent 供应慢时 Pipeline 可能直接失败。
2. **``post`` 条件的执行顺序是固定的**, 与声明顺序无关:
   ``always -> changed -> fixed -> regression -> aborted -> failure -> success -> unstable
   -> unsuccessful -> cleanup``;``cleanup`` 排在**所有**其它条件之后。
3. **stage 内各段的求值时点**: ``options`` 最先, 然后 ``input`` 暂停, 再进入 ``agent`` 块,
   最后评估 ``when``。三个 ``before*`` 开关按 ``beforeOptions > beforeInput > beforeAgent``
   的优先级把 ``when`` 提前。
4. **stage 结构约束**: 一个 stage 必须有且仅有一个 ``steps`` / ``stages`` / ``parallel`` /
   ``matrix``;``parallel`` 或 ``matrix`` 块**内部**的 stage 不能再嵌套 ``parallel``/``matrix``。
5. **matrix 的两段式求值**: ``axis`` 与 ``exclude`` 在**运行开始前**生成静态 cell 集合;
   而 per-cell 指令(agent/environment/input/options/post/tools/when)在**运行时**求值。
"""

from __future__ import annotations

import itertools

# 官方文档给出的 post 条件执行顺序(逐字顺序, 不可调整)
POST_ORDER = ("always", "changed", "fixed", "regression", "aborted",
              "failure", "success", "unstable", "unsuccessful", "cleanup")

STATUSES = ("success", "failure", "unstable", "aborted")

# stage 级 options 只允许这四类(官方: "stage 级别的 options 只能包含 retry、timeout、
# timestamps 这类步骤, 或与 stage 相关的声明式选项(如 skipDefaultCheckout)")
STAGE_ALLOWED_OPTIONS = ("retry", "timeout", "timestamps", "skipDefaultCheckout")

# 只能在 pipeline 块内出现一次、且不能在 stage 内的指令
PIPELINE_ONLY_ONCE = ("parameters", "triggers")
STAGE_FORBIDDEN_PIPELINE_ONLY = ("disableRestartFromStage",)

AGENT_TYPES = ("any", "none", "label", "node", "docker", "dockerfile", "kubernetes")


class JenkinsError(ValueError):
    """Pipeline 校验 / 运行期错误。"""


# ---------------------------------------------------------------- post

def post_condition_runs(cond: str, status: str, prev_status: str | None) -> bool:
    """单个 ``post`` 条件是否触发。"""
    if cond == "always":
        return True
    if cond == "cleanup":
        return True                      # 无论什么状态都跑, 且排在最后
    if cond == "changed":
        return status != prev_status
    if cond == "fixed":
        return status == "success" and prev_status in ("failure", "unstable")
    if cond == "regression":
        # 本次 failure/unstable/aborted 且上次 success
        return status in ("failure", "unstable", "aborted") and prev_status == "success"
    if cond == "unsuccessful":
        return status != "success"
    return cond == status                # aborted / failure / success / unstable


def run_post(declared: list, status: str, prev_status: str | None) -> list:
    """按**官方固定顺序**返回实际执行的 post 条件(与声明顺序无关)。"""
    if status not in STATUSES:
        raise JenkinsError("未知构建状态: %s" % status)
    return [c for c in POST_ORDER if c in set(declared) and post_condition_runs(c, status, prev_status)]


# ---------------------------------------------------------------- agent / timeout

def times_out(scope: str, alloc_seconds: float, work_seconds: float, limit_seconds: float):
    """返回 (是否超时, 计入 timeout 的秒数)。

    顶层 agent: 分配时间**不计入**;stage 级 agent: 分配时间**计入**。
    """
    if scope not in ("top-level", "stage"):
        raise JenkinsError("scope 只能是 top-level / stage")
    counted = work_seconds if scope == "top-level" else alloc_seconds + work_seconds
    return counted > limit_seconds, counted


def validate_agent(atype: str, at_top_level: bool = True):
    """校验 agent 声明; 返回错误列表(空 = 合法)。"""
    errs = []
    if atype not in AGENT_TYPES:
        errs.append("未知 agent 类型 %s(允许: %s)" % (atype, "/".join(AGENT_TYPES)))
    if atype == "none" and not at_top_level:
        errs.append("agent none 只能出现在 pipeline 顶层")
    return errs


# ---------------------------------------------------------------- stage 求值时点

def stage_eval_order(before_options=False, before_input=False, before_agent=False) -> list:
    """返回 stage 内各段的实际求值顺序。

    基线顺序 ``options -> input -> agent -> when``(官方: options 先调用; input 在
    "应用完 options 之后、进入 agent 块或评估 when 之前"暂停; when 默认在进入 agent
    之后评估)。三个开关把 ``when`` 往前提, 优先级 ``beforeOptions > beforeInput >
    beforeAgent``。
    """
    if before_options:
        return ["when", "options", "input", "agent"]
    if before_input:
        return ["options", "when", "input", "agent"]
    if before_agent:
        return ["options", "input", "when", "agent"]
    return ["options", "input", "agent", "when"]


def when_priority_rank(flags: dict) -> int:
    """返回生效的 before* 开关的优先级权重(数字越大越优先, 0 = 都没开)。"""
    if flags.get("beforeOptions"):
        return 3
    if flags.get("beforeInput"):
        return 2
    if flags.get("beforeAgent"):
        return 1
    return 0


# ---------------------------------------------------------------- stage 结构

def validate_stage_body(body_keys, parent_is_parallel_or_matrix=False) -> list:
    """校验 stage 的"有且仅有一个"约束与 parallel/matrix 的不可嵌套约束。"""
    errs = []
    chosen = [k for k in ("steps", "stages", "parallel", "matrix") if k in body_keys]
    if len(chosen) != 1:
        errs.append("一个 stage 必须有且仅有一个 steps/stages/parallel/matrix, 实际 %d 个: %s"
                    % (len(chosen), chosen or "无"))
    if parent_is_parallel_or_matrix and ("parallel" in body_keys or "matrix" in body_keys):
        errs.append("parallel/matrix 块内的 stage 不能再嵌套 parallel/matrix")
    return errs


def validate_options_scope(scope: str, option_names) -> list:
    errs = []
    for name in option_names:
        if scope == "stage" and name not in STAGE_ALLOWED_OPTIONS:
            errs.append("stage 级 options 不允许 %s(仅限 %s)"
                        % (name, "/".join(STAGE_ALLOWED_OPTIONS)))
        if scope == "stage" and name in STAGE_FORBIDDEN_PIPELINE_ONLY:
            errs.append("%s 不能在 stage 内使用" % name)
    return errs


def validate_once(name: str, occurrences: int, in_stage: bool = False) -> list:
    errs = []
    if name in PIPELINE_ONLY_ONCE:
        if occurrences > 1:
            errs.append("%s 在 pipeline 块内只能出现一次" % name)
        if in_stage:
            errs.append("%s 只能出现在 pipeline 块内" % name)
    return errs


# ---------------------------------------------------------------- parallel / matrix

def run_parallel(stages: list, results: dict, fail_fast=False, global_fail_fast=False):
    """返回 (实际执行的 stage 列表, 被中止的 stage 列表)。

    ``failFast true`` 或 ``options { parallelsAlwaysFailFast() }`` 下, 任一并行 stage
    失败即中止其余未完成的 stage。
    """
    order = list(stages)
    executed, aborted = [], []
    stopped = False
    for s in order:
        if stopped:
            aborted.append(s)
            continue
        executed.append(s)
        if results.get(s) == "failure" and (fail_fast or global_fail_fast):
            stopped = True
    return executed, aborted


def matrix_cells(axes: list, excludes: list | None = None) -> list:
    """按官方 axis/exclude 语义生成静态 cell 集合。

    ``axes``: [(名称, [值...]), ...];``excludes``: [ {轴名: {"values": [...]}} ... ]
    —— 一个 exclude 命中当且仅当它列出的**每个**轴子句都匹配。
    """
    names = [a[0] for a in axes]
    cells = [dict(zip(names, vals)) for vals in itertools.product(*[a[1] for a in axes])]
    for ex in excludes or []:
        keep = []
        for c in cells:
            hit = True
            for axis_name, clause in ex.items():
                vals = clause.get("values")
                notvals = clause.get("notValues")
                if vals is not None and c.get(axis_name) not in vals:
                    hit = False
                if notvals is not None and c.get(axis_name) in notvals:
                    hit = False
            if not hit:
                keep.append(c)
        cells = keep
    return cells


RUNTIME_DIRECTIVES = ("agent", "environment", "input", "options", "post", "tools", "when")
STATIC_DIRECTIVES = ("axes", "excludes")

File: 06-DevOps/02-CI-CD/Jenkins声明式流水线/go/jenkins_pipeline.go
// Jenkins 声明式 Pipeline 语义模型(Go 对照实现)。
//
// 语义依据同 python/jenkins_pipeline.py 头部的官方文档引用。
// 运行: cd go && go run .
package main

import (
	"fmt"
	"strings"
)

// PostOrder 官方给出的 post 条件执行顺序(逐字顺序, 不可调整)。
var PostOrder = []string{"always", "changed", "fixed", "regression", "aborted",
	"failure", "success", "unstable", "unsuccessful", "cleanup"}

var statuses = map[string]bool{"success": true, "failure": true,
	"unstable": true, "aborted": true}

// StageAllowedOptions stage 级 options 的白名单。
var StageAllowedOptions = []string{"retry", "timeout", "timestamps", "skipDefaultCheckout"}

// PostConditionRuns 单个 post 条件是否触发。
func PostConditionRuns(cond, status, prev string) bool {
	switch cond {
	case "always", "cleanup":
		return true
	case "changed":
		return status != prev
	case "fixed":
		return status == "success" && (prev == "failure" || prev == "unstable")
	case "regression":
		return (status == "failure" || status == "unstable" || status == "aborted") &&
			prev == "success"
	case "unsuccessful":
		return status != "success"
	}
	return cond == status // aborted / failure / success / unstable
}

// RunPost 按**官方固定顺序**返回实际执行的 post 条件(与声明顺序无关)。
func RunPost(declared []string, status, prev string) ([]string, error) {
	if !statuses[status] {
		return nil, fmt.Errorf("未知构建状态: %s", status)
	}
	set := map[string]bool{}
	for _, d := range declared {
		set[d] = true
	}
	var out []string
	for _, c := range PostOrder {
		if set[c] && PostConditionRuns(c, status, prev) {
			out = append(out, c)
		}
	}
	return out, nil
}

// TimesOut 返回 (是否超时, 计入 timeout 的秒数)。
// 顶层 agent: 分配时间不计入; stage 级 agent: 分配时间计入。
func TimesOut(scope string, allocSeconds, workSeconds, limitSeconds float64) (bool, float64, error) {
	if scope != "top-level" && scope != "stage" {
		return false, 0, fmt.Errorf("scope 只能是 top-level / stage")
	}
	counted := workSeconds
	if scope == "stage" {
		counted = allocSeconds + workSeconds
	}
	return counted > limitSeconds, counted, nil
}

// StageEvalOrder 返回 stage 内各段的实际求值顺序。
// 基线 options -> input -> agent -> when; 三个开关按 beforeOptions > beforeInput >
// beforeAgent 的优先级把 when 往前提。
func StageEvalOrder(beforeOptions, beforeInput, beforeAgent bool) []string {
	switch {
	case beforeOptions:
		return []string{"when", "options", "input", "agent"}
	case beforeInput:
		return []string{"options", "when", "input", "agent"}
	case beforeAgent:
		return []string{"options", "input", "when", "agent"}
	}
	return []string{"options", "input", "agent", "when"}
}

// WhenPriorityRank 生效的 before* 开关权重(越大越优先)。
func WhenPriorityRank(beforeOptions, beforeInput, beforeAgent bool) int {
	switch {
	case beforeOptions:
		return 3
	case beforeInput:
		return 2
	case beforeAgent:
		return 1
	}
	return 0
}

// ValidateStageBody 校验"有且仅有一个"与 parallel/matrix 不可嵌套。
func ValidateStageBody(hasSteps, hasStages, hasParallel, hasMatrix, parentIsParOrMat bool) []string {
	var errs []string
	n := 0
	for _, b := range []bool{hasSteps, hasStages, hasParallel, hasMatrix} {
		if b {
			n++
		}
	}
	if n != 1 {
		errs = append(errs, fmt.Sprintf(
			"一个 stage 必须有且仅有一个 steps/stages/parallel/matrix, 实际 %d 个", n))
	}
	if parentIsParOrMat && (hasParallel || hasMatrix) {
		errs = append(errs, "parallel/matrix 块内的 stage 不能再嵌套 parallel/matrix")
	}
	return errs
}

// ValidateOptionsScope 校验 options 的作用域白名单。
func ValidateOptionsScope(scope string, names []string) []string {
	var errs []string
	allowed := map[string]bool{}
	for _, a := range StageAllowedOptions {
		allowed[a] = true
	}
	for _, n := range names {
		if scope == "stage" && !allowed[n] {
			errs = append(errs, fmt.Sprintf("stage 级 options 不允许 %s(仅限 %s)",
				n, strings.Join(StageAllowedOptions, "/")))
		}
	}
	return errs
}

// RunParallel 返回 (实际执行的 stage, 被中止的 stage)。
func RunParallel(stages []string, results map[string]string,
	failFast, globalFailFast bool) ([]string, []string) {
	var executed, aborted []string
	stopped := false
	for _, s := range stages {
		if stopped {
			aborted = append(aborted, s)
			continue
		}
		executed = append(executed, s)
		if results[s] == "failure" && (failFast || globalFailFast) {
			stopped = true
		}
	}
	return executed, aborted
}

// ExcludeClause 是一个 exclude 里的单个轴子句。
type ExcludeClause struct {
	Values    []string
	NotValues []string
}

// MatrixCells 按官方 axis/exclude 语义生成静态 cell 集合。
func MatrixCells(axes [][2]interface{}, excludes []map[string]ExcludeClause) []map[string]string {
	// 展开笛卡尔积
	cells := []map[string]string{{}}
	for _, ax := range axes {
		name := ax[0].(string)
		vals := ax[1].([]string)
		var next []map[string]string
		for _, base := range cells {
			for _, v := range vals {
				c := map[string]string{}
				for k, val := range base {
					c[k] = val
				}
				c[name] = v
				next = append(next, c)
			}
		}
		cells = next
	}
	for _, ex := range excludes {
		var keep []map[string]string
		for _, c := range cells {
			hit := true
			for axisName, clause := range ex {
				if clause.Values != nil && !contains(clause.Values, c[axisName]) {
					hit = false
				}
				if clause.NotValues != nil && contains(clause.NotValues, c[axisName]) {
					hit = false
				}
			}
			if !hit {
				keep = append(keep, c)
			}
		}
		cells = keep
	}
	return cells
}

func contains(xs []string, v string) bool {
	for _, x := range xs {
		if x == v {
			return true
		}
	}
	return false
}

